DESIGN.md

```diff
@@ -0,0 +1,756 @@
+socket receives data, an exchange that can take an arbitrary amount of time.
+Suppose each read operation requires pre-allocating memory for the in-flight
+operation. In that case, the amount of memory consumed by the process grows
+linearly with the number of in-flight read operations. For applications with a
+large number of open connections, this can be problematic. The io-uring API
+supports registering buffer pools with the ring and configuring read operations
+to use the buffer pool instead of a dedicated per-operation buffer. When the
+socket receives data, the kernel checks out a buffer from the pool and returns
+it to the caller. After reading the data, the caller returns the buffer to the
+kernel. If no buffers are available, the operation pauses until the process
+returns a buffer to the kernel, at which time the operation completes.
+
+```rust
+// ...
+let buf = file.read_at_prepared(0, 1024, DefaultPool).await?;
+
+// Write some data back to the file.
+let BufResult(res, buf) = file.write_at(1024, buf.slice(0..512)).await;
+res?;
+
+// Dropping the buffer lets the kernel know the buffer may
+// be reused.
+drop(buf);
+```
+
+The runtime supports initializing multiple pools containing buffers of different
+sizes. When submitting a read operation, the caller specifies the pool from
+which to draw a buffer.
+
+```rust
+// Allocate a buffer pool
+let my_pool = BufferPool::builder()
+    .buffer_size(16_384)
+    .num_buffers(256)
+    .build();
+
+// Create the runtime
+let mut rt = tokio_uring::runtime::Runtime::new()?;
+
+// Provide the buffer pool to the kernel. This passes
+// ownership of the pool to the kernel.
+let pool_token = rt.provide_buffers(MY_POOL, 256, 4 * 4096)?;
+    
+rt.block_on(async {
+    // ...
+    let buf = file.read_at_prepared(0, 1024, pool_token).await?;
+});
+```
+
+## Buffer management
+
+Buffers passed to read and write operations must remain alive and pinned to a
+memory location while operations are in-flight, ruling out `&mut [u8]` as an
+option. Additionally, read and write operations may reference buffers using a
+pointer or, when the buffer is part of a pre-registered buffer pool, using a
+numeric buffer identifier. 
+
+The tokio-uring crate will provide its buffer type, `IoBuf`, to use when reading
+and writing.
+
+```rust
+pub struct IoBuf {
+    kind: Kind,
+}
+
+enum Kind {
+    /// A vector-backed buffer
+    Vec(Vec<u8>),
+
+    /// Buffer pool backed buffer. The pool is managed by io-uring.
+    Provided(ProvidedBuf),
+}
+```
+
+Internally, the IoBuf type is either backed by individually heap-allocated
+memory or a buffer pool entry. Acquiring an `IoBuf` is done via
+`IoBuf::with_capacity` or checking-out an entry from a pool. 
+
+```rust
+// Individually heap-allocated
+let my_buf = IoBuf::with_capacity(4096);
+
+// Checked-out from a pool
+match my_buffer_pool.checkout() {
+    Ok(io_buf) => ...,
+    Err(e) => panic!("buffer pool empty"),
```

Comment:
If "the buffer pool is empty" is the only possible error here, I'd prefer `Option` over `Result`, since it doesn't really feel like an error condition.

Reply:
Perhaps, there are conflicting patterns in std. I don' have a huge opinion there.

Reply:
My general observation has been that `Option` should be used if the alternative case a) carries no information; b) is common, not exceptional; and c) is unlikely to later be propagated as an `Err`. I _think_ all of those apply here, and therefore that `Option` is the better choice. The exception would be if you anticipate that `checkout()` can fail for _different_ reasons as well. But even in that case, I'd prefer `Result<Option<Pool>, E>`.